lib/ledger_sync/adaptors/ledger_serializer_attribute.rb

```diff
@@ -0,0 +1,150 @@
+        @serializer = serializer
+        @type = type.new
+      end
+
+      def block_value_for(resource:)
+        block.call(resource)
+      end
+
+      # Make nested/dot calls (e.g. `vendor.ledger_id`)
+      def dot_value_for(resource:)
+        resource_attribute_dot_parts.inject(resource) { |r, dot_method| r&.send(dot_method) }
+      end
+
+      def ledger_attribute_dot_parts
+        @ledger_attribute_dot_parts ||= ledger_attribute.split('.')
+      end
+
+      def ledger_attribute_hash_for(resource:)
+        ledger_attribute_hash_with(value: value_from_local(resource: resource))
+      end
+
+      # Create nested hash for ledger
+      #
+      # when ledger_attribute = 'Foo.Bar.Baz',
+      # ledger_attribute_hash_with(value: 123)
+      # Result:
+      # {
+      #   'Foo' => {
+      #     'Bar' => {
+      #       'Baz' => 123
+      #     }
+      #   }
+      # }
+      def ledger_attribute_hash_with(value:)
+        ledger_attribute_dot_parts.reverse.inject(value) { |a, n| { n => a } }
+      end
+
+      def references_many?
+        type.is_a?(LedgerSerializerType::ReferencesMany)
+      end
+
+      def resource_attribute?
+        return false if resource_attribute.nil?
+        return false if resource_attribute == ''
+
+        true
+      end
+
+      def resource_attribute_dot_parts
+        @resource_attribute_dot_parts ||= resource_attribute.split('.')
+      end
+
+      def value_from_local(resource:)
+        value = if block.nil?
+                  dot_value_for(resource: resource)
+                else
+                  block_value_for(resource: resource)
+                end
+
+        if references_many?
+          type.convert_from_local(
+            serializer: serializer,
+            value: value
+          )
+        else
+          type.convert_from_local(
+            value: value
+          )
+        end
+      end
+
+      def value_from_ledger(hash:, resource:)
+        value = hash.dig(*ledger_attribute_dot_parts)
+
+        value = if references_many?
+                  type.convert_from_ledger(
+                    resource_class: resource_class,
+                    serializer: serializer,
+                    value: value
+                  )
+                else
+                  type.convert_from_ledger(
+                    value: value
+                  )
+                end
+
+        return value if resource_attribute_dot_parts.count <= 1
+
+        nested_resource = resource.send(resource_attribute_dot_parts.first)&.dup
+        nested_resource ||= resource_attribute_class(resource: resource).new
+
+        build_resource_value_from_nested_attributes(
+          nested_resource,
+          value,
+          resource_attribute_dot_parts[1..-1]
+        )
+      end
+
+      private
+
+      def build_resource_value_from_nested_attributes(resource, value, attribute_parts)
+        first_attribute, *remaining_attributes = attribute_parts
```

Comment:
Maybe the `dup` from line 113 should happen in here, seems safer in case the caller of this method forgets to `dup` the resource